docker.go

```diff
@@ -628,26 +628,38 @@ func (c *DockerContainer) StartLogProducer(ctx context.Context) error {
 			Since:      since,
 		}
 
-		ctx, cancel := context.WithTimeout(ctx, time.Second*5)
-		defer cancel()
-
 		r, err := c.provider.client.ContainerLogs(ctx, c.GetContainerID(), options)
 		if err != nil {
-			// if we can't get the logs, panic, we can't return an error to anything
-			// from within this goroutine
-			panic(err)
+			// if we can't get the logs, retry in one second.
+			if ctx.Err() != nil {
+				// context done.
+				return
+			}
+
+			c.logger.Printf("cannot get logs for container %q: %v", c.ID, err)
+
+			select {
+			case <-ctx.Done():
+				return
+			case <-time.After(1 * time.Second):
+			}
+
+			goto BEGIN
 		}
 		defer c.provider.Close()
 
 		for {
 			select {
-			case <-stop:
-				err := r.Close()
-				if err != nil {
-					// we can't close the read closer, this should never happen
-					panic(err)
+			case <-ctx.Done():
+				if ctx.Err() != nil {
+					err := r.Close()
+					if err != nil {
+						// we can't close the read closer, this should never happen
+						panic(err)
```

Comment:
Why are we panicking here? If you just want to log the error and return then isn't it better to consistently do it in the same way? Regarding logging, I see both usage of `c.logger.Printf` and `fmt.Fprintf(os.Stderr,` which seems strange.

If the purpose is to print the stack trace then maybe panic is best, but why here?